nexus/pkg/artifacts/saver.go

```diff
@@ -68,187 +53,309 @@ func (as *ArtifactSaver) createArtifact() (string, *string) {
-				ArtifactID:         artifactID,
-				Name:               entry.Path,
-				Md5:                md5Checksum,
-				ArtifactManifestID: &manifestID,
-			},
-		)
+	// Prepare all file specs.
+	fileSpecs := []gql.CreateArtifactFileSpecInput{}
+	for name, entry := range manifest.Contents {
+		if entry.LocalPath == nil {
+			continue
+		}
+		fileSpec := gql.CreateArtifactFileSpecInput{
+			ArtifactID:         artifactID,
+			Name:               name,
+			Md5:                entry.Digest,
+			ArtifactManifestID: &manifestID,
+		}
+		fileSpecs = append(fileSpecs, fileSpec)
 	}
-	response, err := gql.CreateArtifactFiles(
-		as.Ctx,
-		as.GraphqlClient,
-		gql.ArtifactStorageLayoutV2,
-		artifactFiles,
-	)
-	if err != nil {
-		err = fmt.Errorf("CreateArtifactFiles: %s, error: %+v response: %+v", as.Artifact.Name, err, response)
-		as.Logger.CaptureFatalAndPanic("sendManifestFiles", err)
-	}
-	for n, edge := range response.GetCreateArtifactFiles().GetFiles().Edges {
-		task := uploader.UploadTask{
-			Url:                *edge.Node.GetUploadUrl(),
-			Path:               man.Contents[n].LocalPath,
-			CompletionCallback: func(*uploader.UploadTask) { as.WgOutstanding.Done() },
+
+	// Upload in batches.
+	numInProgress, numDone := 0, 0
+	nameToScheduledTime := map[string]time.Time{}
+	taskResultsChan := make(chan TaskResult)
+	fileSpecsBatch := make([]gql.CreateArtifactFileSpecInput, 0, batchSize)
+	for numDone < len(fileSpecs) {
+		// Prepare a batch.
+		now := time.Now()
+		fileSpecsBatch = fileSpecsBatch[:0]
+		for _, fileSpec := range fileSpecs {
+			if _, ok := nameToScheduledTime[fileSpec.Name]; ok {
+				continue
+			}
+			nameToScheduledTime[fileSpec.Name] = now
+			fileSpecsBatch = append(fileSpecsBatch, fileSpec)
+			if len(fileSpecsBatch) >= batchSize {
+				break
+			}
+		}
+		if len(fileSpecsBatch) > 0 {
+			// Fetch upload URLs.
+			response, err := gql.CreateArtifactFiles(
+				as.Ctx,
+				as.GraphqlClient,
+				fileSpecsBatch,
+				gql.ArtifactStorageLayoutV2,
+			)
+			if err != nil {
+				return err
+			}
+			if len(fileSpecsBatch) != len(response.CreateArtifactFiles.Files.Edges) {
+				return fmt.Errorf(
+					"expected %v upload URLs, got %v",
+					len(fileSpecsBatch),
+					len(response.CreateArtifactFiles.Files.Edges),
+				)
+			}
+			// Save birth artifact ids, schedule uploads.
+			for i, edge := range response.CreateArtifactFiles.Files.Edges {
+				name := fileSpecsBatch[i].Name
+				entry := manifest.Contents[name]
+				entry.BirthArtifactID = &edge.Node.Artifact.Id
+				manifest.Contents[name] = entry
+				if edge.Node.UploadUrl == nil {
+					numDone++
+					continue
+				}
+				numInProgress++
+				task := &uploader.UploadTask{
+					Path:    *entry.LocalPath,
+					Url:     *edge.Node.UploadUrl,
+					Headers: edge.Node.UploadHeaders,
+					CompletionCallback: func(task *uploader.UploadTask) {
+						taskResultsChan <- TaskResult{task, name}
+					},
+				}
+				as.UploadManager.AddTask(task)
```

Comment:
FYI: i think we still need to add something (in a later PR) to capture uploaded files so we can pass them in filestream as completed files..  See "Uploaded" in pkg/filestream/loop_transmit.go -- and equiv logic in python internal/filestream.go